File: architecture-skills/design-patterns/examples/prototype/prototype.swift
// Prototype pattern in Swift.
//
// Value types (structs) copy on assignment, so for them the pattern is a
// language feature — shown first. Reference types (classes) need an
// explicit clone() that copies mutable state, shown second.

import Foundation

// --- Variant 1: struct — copying is built in ---

struct Circle {
    var x: Double
    var y: Double
    let radius: Double
    let color: String
    var tags: [String]

    func describe() -> String {
        "Circle r=\(Int(radius)) \(color) at (\(Int(x)), \(Int(y))) tags=\(tags)"
    }
}

// --- Variant 2: class — clone() must be explicit ---

final class CircleRef {
    var x: Double
    var y: Double
    let radius: Double
    let color: String
    var tags: [String]

    init(x: Double, y: Double, radius: Double, color: String, tags: [String]) {
        self.x = x
        self.y = y
        self.radius = radius
        self.color = color
        self.tags = tags
    }

    func clone() -> CircleRef {
        // Array is a value type, so assigning it here already copies it.
        CircleRef(x: x, y: y, radius: radius, color: color, tags: tags)
    }

    func describe() -> String {
        "CircleRef r=\(Int(radius)) \(color) at (\(Int(x)), \(Int(y))) tags=\(tags)"
    }
}

let prototypeMarker = Circle(x: 0, y: 0, radius: 12, color: "red", tags: ["alert"])
var movedMarker = prototypeMarker // struct assignment IS the clone
movedMarker.x = 40
movedMarker.y = 25
movedMarker.tags.append("muted")

print(prototypeMarker.describe())
print(movedMarker.describe())

let refPrototype = CircleRef(x: 0, y: 0, radius: 12, color: "red", tags: ["alert"])
let refCopy = refPrototype.clone()
refCopy.x = 40
refCopy.tags.append("muted")

print(refPrototype.describe())
print(refCopy.describe())

File: architecture-skills/design-patterns/examples/prototype/prototype.py
"""Prototype pattern in Python.

Note: the pattern is built into the standard library as the `copy` module.
`copy.deepcopy` handles nested state; customize with __copy__/__deepcopy__
when some fields must be shared or reset instead of duplicated.
"""

from __future__ import annotations

import copy
from dataclasses import dataclass, field


@dataclass
class Circle:
    x: float
    y: float
    radius: float
    color: str
    tags: list[str] = field(default_factory=list)

    def clone(self) -> Circle:
        # deepcopy so the copy gets its own tags list, not an alias
        return copy.deepcopy(self)

    def describe(self) -> str:
        return f"Circle r={self.radius} {self.color} at ({self.x}, {self.y}) tags={self.tags}"


class ShapeRegistry:
    """Catalog of pre-configured prototypes, cloned instead of constructed."""

    def __init__(self) -> None:
        self._prototypes: dict[str, Circle] = {}

    def register(self, name: str, prototype: Circle) -> None:
        self._prototypes[name] = prototype

    def create(self, name: str) -> Circle:
        return self._prototypes[name].clone()


if __name__ == "__main__":
    registry = ShapeRegistry()
    registry.register(
        "warning-marker", Circle(x=0, y=0, radius=12, color="red", tags=["alert"])
    )

    first = registry.create("warning-marker")
    second = registry.create("warning-marker")
    second.x, second.y = 40, 25
    second.tags.append("muted")

    print(first.describe())
    print(second.describe())
    print("independent copies:", first.tags is not second.tags)
